lib/pushmi_pullyu/cli.rb

```diff
@@ -18,26 +18,47 @@ def initialize
   def parse(args = ARGV)
     parse_options(args)
     parse_commands(args)
-    setup_log
   end
 
   def run
-    # Trap interrupts to quit cleanly.
-    Signal.trap('INT') { abort }
-
-    print_banner
-
     if config.daemonize
-      start_working_loop_in_daemon
+      start_server_as_daemon
     else
       # If we're running in the foreground sync the output.
       $stdout.sync = $stderr.sync = true
-      start_working_loop
+      start_server
+    end
+  end
+
+  def start_server
+    setup_signal_traps
+    setup_log
+    print_banner
+
+    begin
+      run_tick_loop
+    rescue Interrupt
+      logger.info 'Shutting down'
+      @running = false
+      logger.info 'Bye!'
+      exit(0)
     end
   end
 
   private
 
+  def setup_signal_traps
+    Signal.trap('INT') { raise Interrupt }
+    Signal.trap('TERM') { raise Interrupt }
+    Signal.trap('HUP') do
```

Comment:
I'll merge as is...but still figuring out how to reopen/rotate logs on SIGHUPs. So ill add another PR for this. As this one is getting pretty beefy as is